aten/src/ATen/native/cuda/TensorShape.cu

```diff
@@ -0,0 +1,385 @@
+  int64_t num_blocks = 0;
+  for (const auto& split_chunk_size : split_chunk_sizes) {
+    num_blocks +=
+        detail::div_up(split_chunk_size, BLOCK_SIZE * BYTES_PER_THREAD);
+  }
+
+  // Calculate the maximum number of blocks to launch. Only consider
+  // maxThreadsPerMultiProcessor as a limiting factor as the kernel uses no
+  // shared memory and little registers. Over-subscribe the SMs to hide I/O
+  // latency.
+  const auto num_sms =
+      at::cuda::getCurrentDeviceProperties()->multiProcessorCount;
+  const auto max_threads_per_sm =
+      at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor;
+  const int64_t max_blocks = num_sms * max_threads_per_sm / BLOCK_SIZE * 2.0;
+
+  // Make each thread process BYTES_PER_THREAD * iter_factor bytes to regulate
+  // block size. Spread iter_factor evenly between chunks_per_block and
+  // iters_per_chunk.
+  int64_t iter_factor = detail::div_up(num_blocks * num_chunks, max_blocks);
+  int64_t chunks_per_block = std::ceil(std::sqrt(iter_factor));
+  chunks_per_block = std::min(chunks_per_block, num_chunks);
+  const int64_t iters_per_chunk = detail::div_up(iter_factor, chunks_per_block);
+
+  // Launch a logically jagged grid of shape
+  // (chunk_size*, num_splits, num_chunks / chunks_per_block)
+  // backed by a physical grid of shape
+  // (sum(chunk_size), num_chunks / chunks_per_block).
+  // A block can find its split_idx via block_idx_to_split_idx.
+  std::vector<int64_t> block_idx_to_split_idx;
+  std::vector<int64_t> blocks_cumsums{0};
+  block_idx_to_split_idx.reserve(num_blocks);
+  for (size_t split_idx = 0; split_idx < split_sizes.size(); ++split_idx) {
+    const auto blocks = detail::div_up(
+        split_chunk_sizes[split_idx],
+        BLOCK_SIZE * BYTES_PER_THREAD * iters_per_chunk);
+    block_idx_to_split_idx.insert(
+        block_idx_to_split_idx.end(), blocks, split_idx);
+    blocks_cumsums.push_back(blocks_cumsums.back() + blocks);
+  }
+
+  dim3 blocks(blocks_cumsums.back(), num_chunks / chunks_per_block, 1);
+  dim3 threads(BLOCK_SIZE, 1, 1);
+
+  auto [_, ptrs] = detail::pack_vecs(
+      {&dst_base_addrs,
+       &src_base_addrs,
+       &split_chunk_sizes,
+       &block_idx_to_split_idx,
+       &blocks_cumsums},
+      device);
+
+  detail::split_with_sizes_copy_out_contiguous_no_cast_kernel<<<
+      blocks,
+      threads,
+      0,
+      at::cuda::getCurrentCUDAStream()>>>(
+      /*dst_base_addrs=*/reinterpret_cast<char**>(ptrs[0]),
+      /*src_base_addrs=*/reinterpret_cast<char**>(ptrs[1]),
+      /*split_chunk_sizes=*/ptrs[2],
+      /*block_idx_to_split_idx=*/ptrs[3],
+      /*blocks_cumsums=*/ptrs[4],
+      src_stride,
+      num_chunks);
+  C10_CUDA_KERNEL_LAUNCH_CHECK();
+
+  for (const auto& t : out) {
+    t.unsafeGetTensorImpl()->bump_version();
+  }
+}
+
+void split_with_sizes_copy_out_cuda(
+    const Tensor& self,
+    IntArrayRef split_sizes,
+    int64_t dim,
+    TensorList out) {
```

Comment:
Could you move the shape checking for the content of out that you do above in this function. This check should be done for the slowpath as well.
Note that if we want to support the full out= variant semantic (which the current kernel doesn't and you don't need it for your use case, so it is fine if you don't do it), we should at::native::resize_output(...) to make each output the right size instead of erroring out..

Reply:
Interesting! This auto-resizing is actually **highly desirable** for our use case. Without it, each output must be reshaped outside the op, which incurs `num_param` extra view ops. I errored out on shape-mismatch because I thought we had to match the semantic of `.copy_` which requires matching shapes. It's fantastic to know that we can auto-reshape here!

Reply:
@albanD I'd love to auto-resize the splits inside the kernel, however the current `CompositeExplicitAutograd` impl doesn't do it.

```python
>>> foo = torch.rand(10, 5)
>>> torch.split_with_sizes_copy(foo, [5, 5], dim=0, out=[torch.empty(5, 5), torch.empty(5, 5)])
>>> torch.split_with_sizes_copy(foo, [5, 5], dim=0, out=[torch.empty(5, 5), torch.empty(25)])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: The size of tensor a (25) must match the size of tensor b (5) at non-singleton dimension 1
```

Should we (1) keep the shape check to match with the current semantic or (2) update the `CompositeExplicitAutograd` impl to support auto-reshaping?

Another potential option is to introduce a flag for reshaping on the out-variant of the view-op. I don't think it's possible today becasuse the flag only makes sense for the out-variants but I don't think the codegen supports out-variant-only flags.

Reply:
> however the current CompositeExplicitAutograd impl doesn't do it.

Ho yeah that's what I said above, the current impl is not properly following the out= semantic that it should.